### DISCUSSION TITLE
Livewire Components Manifest won't update after the component renamed

### DISCUSSION THREAD

**Discussion by ousid** [livewire/livewire#5391] [1. Help] [answered]
After few tests, I noticed when you are trying to fetch the component list via the `component finder manifest`, won't get updated, once you __updated__ or __renamed__ the component, unless you ran `php artisan livewire:discover`

Is it a cache issue, or this behavior is normal?

### Steps to produce
If you want to fetch components list:

```php
app(\Livewire\LivewireComponentsFinder::class)->getManifest();
```

Once you create for example a `CreateUser` component:
```bash
php artisan make:livewire create-user
```

the `manifest` returns the following:
```php
array(
'create-user' => 'App\Http\Livewire\CreateUser'
);
```

But when you rename the component to something like `edit-user`, the manifest still the same, until you run the `livewire:discover` command.

Is this approach logical, or I'm missing something?

Thanks!

**Comment by joshhanley** [comment, marked as answer]
@ousid how are you renaming the component?

**Reply by ousid** [reply]
@joshhanley via `php artisan livewire:mv ...`

**Reply by joshhanley** [reply]
@ousid ah ok, hmm strange, I would have thought it might update the manifest.

But I wouldn't rely on the manifest, as it can be out of date. If a component isn't found in the manifest, then Livewire will go and look for it in all the appropriate locations and then update the manifest after it has been found.

Hope this helps!

**Reply by ousid** [reply]
@joshhanley Thanks for your reply!

The thing is, I'm building a relying on the name of the component, and passing it as a slug in the URL, to fetch it.

It's fine, when the component is not found, because I can run the Laravel discovery command.

What I don't get is. Is this behavior a normal thing, or it's a bug of some sort?

**Reply by joshhanley** [reply]
@ousid ah ok, well FYI the manifest isn't considered part of the public API, so use at your own risk (it's probably going to be removed in V3).

As such I would consider the behaviour normal, as the manifest should really only be used internally by Livewire and Livewire has mechanisms to update the manifest at runtime.

Personally, I would just make use of dynamic Livewire components and pass the name to that instead so Livewire handles the finding of the component for you, see PR #2502.

**Reply by ousid** [reply]
@joshhanley Interesting! I saw from the PR's tests, that It will throw `ComponentAttributeMissingOnDynamicComponentException` exception, but I'm not sure how to catch that in term on a Laravel controller.

To be more specific, here's how I'm doing it with the manifest:

```php
    public function __invoke(string $component)
    {
        if (! $this->checkLivewireExistence($component)) {
            abort(404);
        }

        return view('live', [
            'componentName' => $component,
        ]);
    }

    protected function checkLivewireExistence(string $component)
    {
        $manifest = app(LivewireComponentsFinder::class)->getManifest();

        if ($component instanceof Livewire) {
            return (bool) array_search($component, $manifest);
        }

        if (is_string($component)) {
            return (bool) array_key_exists($component, $manifest);
        }

        return false;
    }
```
Thanks in advance, Josh!

**Reply by joshhanley** [reply]
@ousid ah interesting, hmm try using this instead:

```php
Livewire::getClass($componentName);
```

See it's details here, and it throws a component not found exception so you can use a try/catch in your controller

https://github.com/livewire/livewire/blob/0edabd715f0ea241a883694669185f41c54567df/src/LivewireManager.php#L56-L80

Hope this helps!

**Reply by ousid** [reply]
Didn't know about it before! Will do... Thanks man!

**Reply by joshhanley** [reply]
No worries! 🙂
